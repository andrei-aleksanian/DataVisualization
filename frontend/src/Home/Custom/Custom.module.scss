.index {
  display: grid;
  grid:
    'settings   visualization   visualization  visualization'
    'settings   visualization   visualization  visualization'
    'settings   visualization   visualization  visualization';
  grid-template-columns: repeat(4, 1fr);
  height: 100vh;
  overflow: hidden;
}
