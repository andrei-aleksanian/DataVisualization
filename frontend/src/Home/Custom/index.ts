export { default } from './Custom';
export * from './Custom';
