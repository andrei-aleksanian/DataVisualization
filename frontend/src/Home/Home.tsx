import { LinkHero } from 'Components/Link';
import classes from './Home.module.scss';

export const TEXT_LINK_LIBRARY = 'See an Example';
export const TEXT_LINK_CUSTOM_DATA = 'Visualize Your Data';

export const TEXT_H1 = 'COVA/ANGEL Visualization';
export const TEXT_P =
  "Dimensionality reduction for cohort data visualization is to plot a high-dimensional dataset in 2D,3D spaces to help analysts identify data properties and gain insights visually. COVA and ANGEL are two novel data visualization algorithms that provide a simple but effective way to improve cohort data visualization performance comprehensively. This demo website illustrates examples embedded by these two algorithms under different hyper-parameter settings and supports users in visualizing their datasets. In addition, the fast-ANGEL version is applied to the users' custom datasets. The current restriction of input data format is '.mat'.";
const Home = () => {
  return (
    <div className={classes.index}>
      <div className={classes.Container}>
        <h1>{TEXT_H1}</h1>
        <p>{TEXT_P}</p>
        <div className={classes.LinkWrapper}>
          <LinkHero text={TEXT_LINK_LIBRARY} link="/examples" />
          <LinkHero text={TEXT_LINK_CUSTOM_DATA} link="/your-data" />
        </div>
      </div>
    </div>
  );
};

export default Home;
