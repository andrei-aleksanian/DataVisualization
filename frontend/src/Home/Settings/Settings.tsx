import Slider from 'Components/Forms/Slider';
import { useState } from 'react';
import { LinkBack } from 'Components/Link';

import { Algorithm } from 'types/Settings';
import CheckBoxes from 'Components/Forms/CheckBoxes';

import classes from './Settings.module.scss';
import COVA, { defaultSettingsCOVA } from './components/COVA';
import ANGEL, { defaultSettingsANGEL } from './components/ANGEL';

export const TEXT_CHECKBOX_COVA = 'COVA';
export const TEXT_CHECKBOX_ANGEL = 'ANGEL';
export const TEXT_H1 = 'Example: Cylinder';
export const TEXT_CHECKBOX_ALGORITHM = 'Algorithm:';
export const TEXT_SLIDER_NEIGHBOUR = 'Neighbour:';
export const TEXT_SLIDER_LAMBDA = 'Lambda:';
export const TEXT_CHECKBOX_PRESERVATION = 'Show Data Preservation:';

export enum DataPreservation {
  ON,
  OFF,
}
export interface SettingsCommon {
  dataPreservation: DataPreservation;
  algorithm: Algorithm;
  neighbour: number;
  lambda: number;
}
export const defaultSettingsCommon = {
  dataPreservation: DataPreservation.OFF,
  algorithm: Algorithm.COVA,
  neighbour: 0,
  lambda: 0,
};

const Settings = () => {
  const [settingsCommon, setSettingsCommon] = useState(defaultSettingsCommon);
  const [settingsCOVA, setSettingsCOVA] = useState(defaultSettingsCOVA);
  const [settingsANGEL, setSettingsANGEL] = useState(defaultSettingsANGEL);

  const onChangeAlgorithm = (event: React.ChangeEvent, newAlgorithm: Algorithm) => {
    event.preventDefault();
    setSettingsCommon((prev) => ({ ...prev, algorithm: newAlgorithm }));
  };
  const onChangeNeighbour = (value: number) => setSettingsCommon((prev) => ({ ...prev, neighbour: value }));
  const onChangeLambda = (value: number) => setSettingsCommon((prev) => ({ ...prev, lambda: value }));
  const onChangeDataPreservation = (event: React.ChangeEvent, value: DataPreservation) => {
    event.preventDefault();
    setSettingsCommon((prev) => ({ ...prev, dataPreservation: value }));
  };

  return (
    <div className={classes.index}>
      <LinkBack link="/" />
      <h1>{TEXT_H1}</h1>
      <CheckBoxes
        heading={TEXT_CHECKBOX_ALGORITHM}
        currentValue={settingsCommon.algorithm}
        onChange={onChangeAlgorithm}
        entries={[
          { value: Algorithm.COVA, text: TEXT_CHECKBOX_COVA },
          { value: Algorithm.ANGEL, text: TEXT_CHECKBOX_ANGEL },
        ]}
      />
      <Slider
        min={0}
        max={7}
        step={1}
        marksArr={['10', '20', '30', '10%', '20%', '30%', '40%', '50%']}
        onChange={onChangeNeighbour}
        text={TEXT_SLIDER_NEIGHBOUR}
        value={settingsCommon.neighbour}
      />
      <Slider
        min={0}
        max={1}
        step={0.2}
        marksArr={[0, 0.2, 0.4, 0.6, 0.8, 1]}
        onChange={onChangeLambda}
        text={TEXT_SLIDER_LAMBDA}
        value={settingsCommon.lambda}
      />
      {settingsCommon.algorithm === Algorithm.COVA ? (
        <COVA {...{ settingsCOVA, setSettingsCOVA }} />
      ) : (
        <ANGEL {...{ settingsANGEL, setSettingsANGEL }} />
      )}
      <CheckBoxes
        heading={TEXT_CHECKBOX_PRESERVATION}
        currentValue={settingsCommon.dataPreservation}
        onChange={onChangeDataPreservation}
        entries={[
          { value: DataPreservation.ON, text: 'Yes' },
          { value: DataPreservation.OFF, text: 'No' },
        ]}
      />
    </div>
  );
};

export default Settings;
