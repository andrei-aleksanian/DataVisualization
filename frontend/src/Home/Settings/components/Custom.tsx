import FileDropArea, { FileArea } from 'Components/Forms/FileDropArea';
import Button from 'Components/Forms/Button';
import CheckBoxes from 'Components/Forms/CheckBoxes';
import Error from 'Components/Forms/Error';

import classes from '../Settings.module.scss';

export enum Dimension {
  D2,
  D3,
}
export interface SettingsCustom {
  dimension: Dimension;
  file: FileArea;
}
export const defaultSettingsCustom: SettingsCustom = {
  file: { file: null, error: null },
  dimension: Dimension.D2,
};
export interface CustomProps {
  onSubmit: Function;
  setSettingsCustom: React.Dispatch<React.SetStateAction<SettingsCustom>>;
  settingsCustom: SettingsCustom;
  error: string | null;
  acceptedType: string;
}

export const TEXT_CHECKBOX_DIMENSION = 'Dimension:';
export const TEXT_BUTTON = 'Submit';

const Custom = ({
  onSubmit,
  setSettingsCustom,
  settingsCustom: { dimension, file },
  error,
  acceptedType,
}: CustomProps) => {
  const onChangeDimension = (event: React.ChangeEvent, value: Dimension) => {
    event.preventDefault();
    setSettingsCustom((prev) => ({ ...prev, dimension: value }));
  };
  const setFile = (fileNew: FileArea) => {
    setSettingsCustom((prev) => ({ ...prev, file: fileNew }));
  };

  return (
    <div>
      <CheckBoxes
        heading={TEXT_CHECKBOX_DIMENSION}
        currentValue={dimension}
        onChange={onChangeDimension}
        entries={[
          { value: Dimension.D2, text: '2' },
          { value: Dimension.D3, text: '3' },
        ]}
      />
      <FileDropArea setFile={setFile} file={file} acceptedType={acceptedType} />
      <Button text={TEXT_BUTTON} onClick={() => onSubmit()} active={false} customClass={classes.submit} />
      {error && <Error text={error} />}
    </div>
  );
};

export default Custom;
