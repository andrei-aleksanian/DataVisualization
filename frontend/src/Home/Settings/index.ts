export { default } from './Settings';
export * from './Settings';
export * from './components/ANGEL';
export * from './components/COVA';
