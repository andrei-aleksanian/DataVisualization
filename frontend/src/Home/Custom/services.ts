import axios from 'utils/axios';
import { ParamsCOVA } from 'types/Data/Params';
import { Dimension } from 'Home/Settings/components/Custom';
import { DataPerseveranceLabelled, DATA_PERSEVERANCE } from '../../types/Data/DataPerseverance';

export const getCovaDynamicInit = async (params: ParamsCOVA, file: File, dimension: Dimension) => {
  const formData = new FormData();
  formData.append('dimension', dimension === Dimension.D2 ? '2' : '3');
  formData.append('neighbourNumber', params.neighbourNumber);
  formData.append('lambdaParam', params.lambdaParam.toString());
  formData.append('alpha', params.alpha.toString());
  formData.append('isCohortNumberOriginal', params.isCohortNumberOriginal.toString());
  formData.append('file', file);

  try {
    const config = {
      headers: { 'content-type': 'multipart/form-data' },
    };
    const { data }: { data: DataPerseveranceLabelled } = await axios.post('/dynamic/cova', formData, config);
    return data;
  } catch (e) {
    // eslint-disable-next-line no-console
    console.log(e);
    // todo: return error and handle it
    return DATA_PERSEVERANCE;
  }
};

export const getCovaDynamic = async (body: DataPerseveranceLabelled) => {
  try {
    const { data }: { data: DataPerseveranceLabelled } = await axios.post('/cova-demo-dynamic', body);
    return data;
  } catch (e) {
    // eslint-disable-next-line no-console
    console.log(e);
    // todo: return error and handle it
    return DATA_PERSEVERANCE;
  }
};
