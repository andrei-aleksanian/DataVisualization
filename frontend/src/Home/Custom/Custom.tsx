import { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { DataPerseveranceLabelled, DataPerseveranceColored } from 'types/Data/DataPerseverance';
import { Popup } from 'Components/UI';
import { ParamsANGEL, ParamsCOVA } from 'types/Data/Params';
import { Algorithm } from 'types/Settings';
import { Point2D, Point3D } from 'types/Data';
import getColors from 'utils/getColors';
import { Dimension, SettingsCustom } from 'Home/Settings/components/Custom';
import Settings, {
  defaultSettingsCOVA,
  defaultSettingsANGEL,
  defaultSettingsCommon,
  NEIGHBOUR_MARKS_ARR,
  DataPreservation,
  CohortNumber,
  AnchorModification,
} from '../Settings';
import Visualization2D from '../Visualization2D';

import { getCovaDynamicInit, getCovaDynamic } from './services';

import classes from './Custom.module.scss';

export const defaultSettingsCustom: SettingsCustom = {
  file: null,
  dimension: Dimension.D2,
  validation: {
    validation: false,
  },
};

const Custom = () => {
  const [data, setData] = useState<DataPerseveranceColored | null>(null);
  const [error, setError] = useState<string | null>(null);

  const [settingsCommon, setSettingsCommon] = useState(defaultSettingsCommon);
  const [settingsCOVA, setSettingsCOVA] = useState(defaultSettingsCOVA);
  const [settingsANGEL, setSettingsANGEL] = useState(defaultSettingsANGEL);
  const [settingsCustom, setSettingsCustom] = useState(defaultSettingsCustom);

  const history = useHistory();

  const updateData = (newData: DataPerseveranceLabelled) => {
    setData(() => {
      const colors = getColors(newData.labels);
      return {
        ...newData,
        points: newData.points.map((p) => p.map((p2) => p2 * 100) as Point2D | Point3D),
        colors,
      };
    });
  };

  const onSubmit = async () => {
    setSettingsCustom((prev) => ({ ...prev, validation: { validation: settingsCustom.file === null } }));
    if (settingsCustom.file === null) return;

    let newData: DataPerseveranceLabelled;
    if (settingsCommon.algorithm === Algorithm.COVA) {
      const params: ParamsCOVA = {
        neighbourNumber: NEIGHBOUR_MARKS_ARR[settingsCommon.neighbour],
        alpha: settingsCOVA.alpha,
        isCohortNumberOriginal: settingsCOVA.cohortNumber === CohortNumber.ORIGINAL,
      };
      newData = await getCovaDynamicInit(params, settingsCustom.file, settingsCustom.dimension);
      updateData(newData);
      while (newData.iteration < newData.maxIteration) {
        newData = await getCovaDynamic(newData); // eslint-disable-line
        updateData(newData);
      }
    } else {
      const params: ParamsANGEL = {
        neighbourNumber: NEIGHBOUR_MARKS_ARR[settingsCommon.neighbour],
        anchorDensity: settingsANGEL.anchorDensity,
        epsilon: settingsANGEL.epsilon,
        isAnchorModification: settingsANGEL.anchorModification === AnchorModification.ON,
      };
      console.log(params, settingsCustom.file); // eslint-disable-line no-console
    }
  };

  return (
    <div className={classes.index}>
      <Settings
        {...{
          settingsCommon,
          setSettingsCommon,
          settingsCOVA,
          setSettingsCOVA,
          settingsANGEL,
          setSettingsANGEL,
          backLink: '/',
          customDataPage: {
            onSubmit,
            settingsCustom,
            setSettingsCustom,
          },
          reviewer: false,
          name: 'Custom Data',
        }}
      />
      {data && (
        <Visualization2D data={data} showPreservation={settingsCommon.dataPreservation === DataPreservation.ON} />
      )}
      {error && (
        <Popup
          onClick={() => {
            setError(null);
            history.push('/examples');
          }}
        />
      )}
    </div>
  );
};

export default Custom;
