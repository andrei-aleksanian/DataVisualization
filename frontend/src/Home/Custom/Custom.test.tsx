import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import axios from 'utils/axios';

import { DATA_PERSEVERANCE } from 'types/Data/DataPerseverance';

import { TEST_ID } from 'Components/Forms/FileDropArea';
import { TEXT_BUTTON } from 'Home/Settings';
import Custom, { TEXT_FILE_NULL } from '.';

const mockAxios = () => {
  return jest.spyOn(axios, 'post');
};

test('Examples matches snapshot', async () => {
  const { asFragment } = render(<Custom />, { wrapper: MemoryRouter });
  expect(asFragment()).toMatchSnapshot();
});

describe('Examples fetches data correctly', () => {
  test('Success fetch data on click', async () => {
    const mockedAxios = mockAxios();
    mockedAxios.mockResolvedValueOnce({ data: DATA_PERSEVERANCE });
    mockedAxios.mockResolvedValueOnce({ data: { ...DATA_PERSEVERANCE, iteration: DATA_PERSEVERANCE.maxIteration } });
    render(<Custom />, { wrapper: MemoryRouter });

    const file = new File(['hello'], 'hello.mat', { type: '.mat' });
    const fileArea = screen.getByTestId(TEST_ID) as HTMLInputElement;
    await waitFor(() => userEvent.upload(fileArea, file));

    const submit = screen.getByText(TEXT_BUTTON);
    await waitFor(async () => userEvent.click(submit));

    // after 2 iterations of side effects, the canvas should be in the document
    const canvas = await screen.findByTestId('canvas');
    const error = screen.queryByText('custom-error');
    expect(canvas).toBeInTheDocument();
    expect(error).not.toBeInTheDocument();
    // the algorithm must stop when iterations == maxIterations
    expect(mockedAxios).toHaveBeenCalledTimes(2);
  });

  test('Fail no file', async () => {
    const mockedAxios = mockAxios();
    render(<Custom />, { wrapper: MemoryRouter });

    const submit = screen.getByText(TEXT_BUTTON);
    await waitFor(() => userEvent.click(submit));

    // when providing no .mat file, the algorithm must not run
    // and the error message should appear
    // canvas has no data to be renderred on
    const error = await screen.findByText(TEXT_FILE_NULL);
    const canvas = screen.queryByTestId('canvas');

    expect(error).toBeInTheDocument();
    expect(canvas).not.toBeInTheDocument();
    expect(mockedAxios).not.toHaveBeenCalled();
  });

  test('Fail wrong file', async () => {
    const mockedAxios = mockAxios();
    render(<Custom />, { wrapper: MemoryRouter });

    const file = new File(['hello'], 'hello.jpg', { type: '.jpg' });
    const fileArea = screen.getByTestId(TEST_ID) as HTMLInputElement;
    await waitFor(() => userEvent.upload(fileArea, file));

    const submit = screen.getByText(TEXT_BUTTON);
    await waitFor(() => userEvent.click(submit));

    // when providing a non .mat file, the algorithm must not run
    // and the error message should appear
    // canvas has no data to be renderred on
    const error = await screen.findByText(TEXT_FILE_NULL);
    const canvas = screen.queryByTestId('canvas');

    expect(error).toBeInTheDocument();
    expect(canvas).not.toBeInTheDocument();
    expect(mockedAxios).not.toHaveBeenCalled();
  });

  test('Fail something went wrong on the server', async () => {
    const mockedAxios = mockAxios();
    const ERROR_MESSAGE = 'Oops';
    mockedAxios.mockRejectedValueOnce(new Error(ERROR_MESSAGE));
    mockedAxios.mockResolvedValueOnce({ data: DATA_PERSEVERANCE });
    mockedAxios.mockResolvedValueOnce({ data: { ...DATA_PERSEVERANCE, iteration: DATA_PERSEVERANCE.maxIteration } });
    render(<Custom />, { wrapper: MemoryRouter });

    const file = new File(['hello'], 'hello.mat', { type: '.mat' });
    const fileArea = screen.getByTestId(TEST_ID) as HTMLInputElement;
    await waitFor(() => userEvent.upload(fileArea, file));

    const submit = screen.getByText(TEXT_BUTTON);
    await waitFor(() => userEvent.click(submit));

    // when the server throws an error
    // the algorithm should stop and
    // show an error message
    const error = await screen.findByText(ERROR_MESSAGE);
    const canvas = screen.queryByTestId('canvas');

    expect(error).toBeInTheDocument();
    expect(canvas).not.toBeInTheDocument();
    expect(mockedAxios).toHaveBeenCalledTimes(1);
  });
});
