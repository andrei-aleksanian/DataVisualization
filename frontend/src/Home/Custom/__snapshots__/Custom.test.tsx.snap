// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Examples matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="index"
  >
    <div
      class="index"
    >
      <a
        class="index Back"
        href="/"
      >
        <img
          alt="back"
          src="arrow.svg"
        />
      </a>
      <h1>
        Custom Data
      </h1>
      <div
        class="index"
      >
        <p>
          Algorithm:
        </p>
        <div
          class="CheckBoxes"
        >
          <label>
            <input
              type="radio"
            />
            <span
              class="text"
            >
              COVA
            </span>
            <span
              class="checkmark"
            />
          </label>
          <label>
            <input
              checked=""
              type="radio"
            />
            <span
              class="text"
            >
              ANGEL
            </span>
            <span
              class="checkmark"
            />
          </label>
        </div>
      </div>
      <div
        class="index"
      >
        <p>
          Neighbour number:
        </p>
        <div
          class="rc-slider rc-slider-with-marks"
        >
          <div
            class="rc-slider-rail"
            style="height: 4px;"
          />
          <div
            class="rc-slider-track"
            style="height: 4px; left: 0%; width: 0%;"
          />
          <div
            class="rc-slider-step"
          >
            <span
              class="rc-slider-dot rc-slider-dot-active"
              style="height: 10px; width: 10px; bottom: -3px; left: 0%;"
            />
            <span
              class="rc-slider-dot"
              style="border: 2px solid #fff; height: 10px; width: 10px; bottom: -3px; left: 25%;"
            />
            <span
              class="rc-slider-dot"
              style="border: 2px solid #fff; height: 10px; width: 10px; bottom: -3px; left: 50%;"
            />
            <span
              class="rc-slider-dot"
              style="border: 2px solid #fff; height: 10px; width: 10px; bottom: -3px; left: 75%;"
            />
            <span
              class="rc-slider-dot"
              style="border: 2px solid #fff; height: 10px; width: 10px; bottom: -3px; left: 100%;"
            />
          </div>
          <div
            aria-disabled="false"
            aria-valuemax="4"
            aria-valuemin="0"
            aria-valuenow="0"
            class="rc-slider-handle"
            role="slider"
            style="height: 20px; width: 20px; margin-top: -9px; box-shadow: none; left: 0%; transform: translateX(-50%);"
            tabindex="0"
          />
          <div
            class="rc-slider-mark"
          >
            <span
              class="rc-slider-mark-text rc-slider-mark-text-active"
              style="transform: translateX(-50%); left: 0%; font-size: 0.85rem; color: rgb(204, 204, 204);"
            >
              10
            </span>
            <span
              class="rc-slider-mark-text"
              style="transform: translateX(-50%); left: 25%; font-size: 0.85rem; color: rgb(204, 204, 204);"
            >
              20
            </span>
            <span
              class="rc-slider-mark-text"
              style="transform: translateX(-50%); left: 50%; font-size: 0.85rem; color: rgb(204, 204, 204);"
            >
              30
            </span>
            <span
              class="rc-slider-mark-text"
              style="transform: translateX(-50%); left: 75%; font-size: 0.85rem; color: rgb(204, 204, 204);"
            >
              10%
            </span>
            <span
              class="rc-slider-mark-text"
              style="transform: translateX(-50%); left: 100%; font-size: 0.85rem; color: rgb(204, 204, 204);"
            >
              30%
            </span>
          </div>
        </div>
      </div>
      <div
        class="index"
      >
        <p>
          Anchor density:
        </p>
        <div
          class="rc-slider rc-slider-with-marks"
        >
          <div
            class="rc-slider-rail"
            style="height: 4px;"
          />
          <div
            class="rc-slider-track"
            style="height: 4px; left: 0%; width: 0%;"
          />
          <div
            class="rc-slider-step"
          >
            <span
              class="rc-slider-dot rc-slider-dot-active"
              style="height: 10px; width: 10px; bottom: -3px; left: 0%;"
            />
            <span
              class="rc-slider-dot"
              style="border: 2px solid #fff; height: 10px; width: 10px; bottom: -3px; left: 100%;"
            />
          </div>
          <div
            aria-disabled="false"
            aria-valuemax="0.2"
            aria-valuemin="0.1"
            aria-valuenow="0.1"
            class="rc-slider-handle"
            role="slider"
            style="height: 20px; width: 20px; margin-top: -9px; box-shadow: none; left: 0%; transform: translateX(-50%);"
            tabindex="0"
          />
          <div
            class="rc-slider-mark"
          >
            <span
              class="rc-slider-mark-text rc-slider-mark-text-active"
              style="transform: translateX(-50%); left: 0%; font-size: 0.85rem; color: rgb(204, 204, 204);"
            >
              0.1
            </span>
            <span
              class="rc-slider-mark-text"
              style="transform: translateX(-50%); left: 100%; font-size: 0.85rem; color: rgb(204, 204, 204);"
            >
              0.2
            </span>
          </div>
        </div>
      </div>
      <div
        class="index"
      >
        <p>
          Epsilon:
        </p>
        <div
          class="rc-slider rc-slider-with-marks"
        >
          <div
            class="rc-slider-rail"
            style="height: 4px;"
          />
          <div
            class="rc-slider-track"
            style="height: 4px; left: 0%; width: 0%;"
          />
          <div
            class="rc-slider-step"
          >
            <span
              class="rc-slider-dot rc-slider-dot-active"
              style="height: 10px; width: 10px; bottom: -3px; left: 0%;"
            />
            <span
              class="rc-slider-dot"
              style="border: 2px solid #fff; height: 10px; width: 10px; bottom: -3px; left: 100%;"
            />
          </div>
          <div
            aria-disabled="false"
            aria-valuemax="5"
            aria-valuemin="0.1"
            aria-valuenow="0.1"
            class="rc-slider-handle"
            role="slider"
            style="height: 20px; width: 20px; margin-top: -9px; box-shadow: none; left: 0%; transform: translateX(-50%);"
            tabindex="0"
          />
          <div
            class="rc-slider-mark"
          >
            <span
              class="rc-slider-mark-text rc-slider-mark-text-active"
              style="transform: translateX(-50%); left: 0%; font-size: 0.85rem; color: rgb(204, 204, 204);"
            >
              0.1
            </span>
            <span
              class="rc-slider-mark-text"
              style="transform: translateX(-50%); left: 100%; font-size: 0.85rem; color: rgb(204, 204, 204);"
            >
              5
            </span>
          </div>
        </div>
      </div>
      <div
        class="index"
      >
        <p>
          Anchor modification:
        </p>
        <div
          class="CheckBoxes"
        >
          <label>
            <input
              checked=""
              type="radio"
            />
            <span
              class="text"
            >
              0
            </span>
            <span
              class="checkmark"
            />
          </label>
          <label>
            <input
              type="radio"
            />
            <span
              class="text"
            >
              1
            </span>
            <span
              class="checkmark"
            />
          </label>
        </div>
      </div>
      <div
        class="index"
      >
        <p>
          Show Data Preservation Error:
        </p>
        <div
          class="CheckBoxes"
        >
          <label>
            <input
              type="radio"
            />
            <span
              class="text"
            >
              Yes
            </span>
            <span
              class="checkmark"
            />
          </label>
          <label>
            <input
              checked=""
              type="radio"
            />
            <span
              class="text"
            >
              No
            </span>
            <span
              class="checkmark"
            />
          </label>
        </div>
      </div>
      <div>
        <div
          class="index"
        >
          <p>
            Dimension:
          </p>
          <div
            class="CheckBoxes"
          >
            <label>
              <input
                checked=""
                type="radio"
              />
              <span
                class="text"
              >
                2
              </span>
              <span
                class="checkmark"
              />
            </label>
            <label>
              <input
                type="radio"
              />
              <span
                class="text"
              >
                3
              </span>
              <span
                class="checkmark"
              />
            </label>
          </div>
        </div>
        <p
          class="p"
        >
          Choose a File:
        </p>
        <div
          class="Wrapper"
        >
          <div
            class="Container"
            style="border-color: #bbb;"
            tabindex="0"
          >
            <input
              accept=".mat"
              autocomplete="off"
              data-testid="filearea"
              multiple=""
              style="display: none;"
              tabindex="-1"
              type="file"
            />
            <label>
              Drag and Drop or Click HERE to choose file
            </label>
          </div>
        </div>
        <button
          class="Button submit"
          type="button"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
